### PR TITLE
Issue#27 pycsw single microservice

### LINKED ISSUE #27: pyCSW single microservice
pycsw is currently used as a library inside of GeoNode. To reduce the complexity of the code and improve microservice design pattern, to improve scalability and resource management pycsw setup as a single deployment running pycsw. GeoNode already supports external pycsw service by settings.py parameters

### PR DESCRIPTION
**Pull request by mwallschlaeger**
Closes #27 

## 📑 Description
ref #27

added pycsw microservice inspired by pycsw helm chart (https://github.com/geopython/pycsw/tree/master/docker/helm) with adoption to the style of this helm chart.

Implementation:
- nginx container forwarding traffic to pycsw via  `.Values.pycsw.endpoint` url
- mappings.py taken from GeoNode codbase (https://github.com/GeoNode/geonode/blob/master/geonode/catalogue/backends/pycsw_local_mappings.py)
- pycsw.cfg taken from pycsw helm chart

### CODE REVIEW

**Comment by mwallschlaeger** [comment]
waiting for https://github.com/GeoNode/geonode/pull/11319 . meanwhile you can use the following geonode image including the bugfix promoted inside the PR.
```
  image:
    # -- used geonode image
    name: mwall2bitflow/geonode
    # -- tag of used geonode image
    tag: '4.1.x-pycsw-hotfix'
```

**Comment by mwallschlaeger** [comment]
https://github.com/GeoNode/geonode/pull/11319 got merged, so no reason to hold this back anymore

File: deployment/geonode/templates/pycsw/pycsw-cfg.yaml [hunk 1]
```diff
@@ -0,0 +1,70 @@
+{{ if .Values.pycsw.enabled }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}-pycsw-cfg
+  namespace: {{ .Release.Namespace }}
+data:
+  pycsw.cfg: |+
```
Comment (ridoo) [inline deployment/geonode/templates/pycsw/pycsw-cfg.yaml:8]: Should we better include that configuration as is in the values.yml and embed the config via `{{- include "pycsw.config" . | nindent 4 }}`?

File: deployment/geonode/templates/pycsw/pycsw-deploy.yaml [hunk 2]
```diff
@@ -0,0 +1,91 @@
+{{ if .Values.pycsw.enabled }}
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: "{{ include "pycsw_pod_name" . }}"
```
Comment (ridoo) [inline deployment/geonode/templates/pycsw/pycsw-deploy.yaml:5]: #64

File: deployment/geonode/templates/pycsw/pycsw-env.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{{ if .Values.pycsw.enabled }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ include "pycsw_pod_name" . }}-env
```
Comment (ridoo) [inline deployment/geonode/templates/pycsw/pycsw-env.yaml:5]: #64

File: deployment/geonode/templates/pycsw/pycsw-svc.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+{{- if .Values.pycsw.enabled }}
+apiVersion: v1
+kind: Service
+metadata:
+  name: "{{ include "pycsw_pod_name" . }}"
```
Comment (ridoo) [inline deployment/geonode/templates/pycsw/pycsw-svc.yaml:5]: #64

File: docs/pycsw.md [hunk 5]
```diff
@@ -0,0 +1,8 @@
+External pyCSW 
+--------------
+
+GeoNode-k8s external pycsw container. This allows changes to pyCSW with having to change GeoNode codebase and allows scalability of the pyCSW service seperatly. The external pycsw service is by default enabled. To disable set: `.Values.pycsw.enabled = False`
```
Comment (ridoo) [inline docs/pycsw.md:4]: _without_ having to change GeoNode

File: docs/pycsw.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+
+The pycsw configuration (pycsw.cfg) and mappings are defined in the values file: `[ .Values.pycsw.config , .Values.pycsw.mappings ]`
+
+The csw endpoint is default set to the one used within Geonode: `/catalogue/csw`
```
Comment (ridoo) [inline docs/pycsw.md:8]: Would be nice to have a link reference here, e.g. https://docs.pycsw.org/en/2.6.1/configuration.html

File: minikube-values.yaml [hunk 7]
```diff
@@ -30,6 +30,9 @@ geonode:
     email_verification: "False"
     authentication_method: username_email
 
+pycsw:
+  enabled: True
```
Comment (ridoo) [inline minikube-values.yaml:34]: Really the default?

**Review by ridoo — CHANGES_REQUESTED** [review]

**Review by ridoo — CHANGES_REQUESTED** [review]
We do not have to resolve #64 first, but at least should have a look at the way how the config is embedded in the configMap

File: deployment/geonode/templates/pycsw/pycsw-deploy.yaml (hunk 2 above)
Comment (mwallschlaeger) [reply deployment/geonode/templates/pycsw/pycsw-deploy.yaml:5]: lets work on #64 in another PR

File: deployment/geonode/templates/nginx/nginx-conf.yaml [hunk 8]
```diff
@@ -110,6 +110,34 @@ data:
...
+        # compression
+        gzip on;
+        gzip_types
+          text/xml
+          text/plain
+          application/xml
+          application/xml+rss
+          application/json;
+
+        proxy_pass http://{{ include "pycsw_pod_name" . }}:{{ .Values.pycsw.port }};
```
[the remark this answers is not in the document]
Comment (mwallschlaeger) [reply deployment/geonode/templates/nginx/nginx-conf.yaml:133]: lets work on #64 in another PR

File: deployment/geonode/templates/geonode/geonode-env.yaml [hunk 9]
```diff
@@ -127,6 +127,13 @@ data:
   LDAP_USER_ATTR_MAP_LAST_NAME: {{ .Values.geonode.ldap.attr_map_last_name | quote }}
   LDAP_USER_ATTR_MAP_EMAIL_ADDR: {{ .Values.geonode.ldap.attr_map_email_addr | quote }}
 
+  # Configure PYCSW
+  {{ if .Values.pycsw.enabled }}
+  CATALOGUE_ENGINE: geonode.catalogue.backends.pycsw_http
+  CATALOGUE_URL: "http://{{ include "pycsw_pod_name" . }}:{{ .Values.pycsw.port }}"
```
[the remark this answers is not in the document]
Comment (mwallschlaeger) [reply deployment/geonode/templates/geonode/geonode-env.yaml:133]: lets work on #64 in another PR

File: deployment/geonode/templates/_helpers.tpl [hunk 10]
```diff
@@ -17,6 +17,10 @@
 {{ .Release.Name }}-{{ .Values.nginx.pod_name }}
 {{- end -}}
 
+{{- define "pycsw_pod_name" -}}
+{{ .Release.Name }}-{{ .Values.pycsw.pod_name }}
```
[the remark this answers is not in the document]
Comment (mwallschlaeger) [reply deployment/geonode/templates/_helpers.tpl:21]: lets work on #64 in another PR

File: deployment/geonode/templates/pycsw/pycsw-env.yaml (hunk 3 above)
Comment (mwallschlaeger) [reply deployment/geonode/templates/pycsw/pycsw-env.yaml:5]: lets work on #64 in another PR

**Pull request closed without merging**